### ISSUE TITLE
Optimizations like using GPT RETRO / sub-quadratic attention?

### ISSUE THREAD

**Issue by vedantroy** [lucidrains/imagen-pytorch#143]
Hi! I've been working on implementing Imagen from scratch, partially for learning, and partially so I can implement optimizations like sub-quadratic attention & try out things like using GPT RETRO instead of T5. Mostly, I want to see if I can reduce the training cost by 1 OOM.

I have a bit of a ways to go, before I can start implementing those optimizations (currently working on getting something basic running on the CelebHQ dataset), but I'm curious if you think:
- Sub-quadratic attention could work in Imagen
- GPT RETRO could be used to replace T5

I figured I'd ask you, since you seem like an expert in this stuff, and I started learning ML 1 month ago 😅.

**Comment by jorgemcgomes** [comment]
It is a bit puzzling to me why T5 is a crucial component of Imagen. It probably isn't? It's probably just the size and dimension of the encoder that matters. I would imagine that any sufficiently big and highly capable text transformer (encoder) would also work fine.
There have been some experiments on the capabilities of the T5 encoder (https://arxiv.org/abs/2108.08877), and there's nothing very special when compared to other transformers of comparable size. 

You can also see this in the Imagen paper. T5-Base performs similarly to BERT-Large, which have the similar model dimension and comparable sizes.
![image](https://user-images.githubusercontent.com/3987574/180430613-a0de2d07-fd65-418c-af94-9be643be97f3.png)

**Comment by jacobwjs** [comment]
I'm sitting on a bit of data now waiting to test along these very lines. There is nothing in my mind that should fundamentally require T5 (would love to be corrected here otherwise). I think this is more of google using google research (i.e. T5).

And if Imagen scales with capacity in the encoder as @jorgemcgomes linked from the paper then I would assume larger capacity models would move the needle past what we've seen from T5. 

Unfortunately the issue is compute as always. Anytime you'd want to infer something requires quite a heavy load, which is the main reason I haven't pushed forward.

@vedantroy keep us updated if you test this out!

**Comment by lucidrains** [comment]
> I'm sitting on a bit of data now waiting to test along these very lines. There is nothing in my mind that should fundamentally require T5 (would love to be corrected here otherwise). I think this is more of google using google research (i.e. T5).
> 
> And if Imagen scales with capacity in the encoder as @jorgemcgomes linked from the paper then I would assume larger capacity models would move the needle past what we've seen from T5.
> 
> Unfortunately the issue is compute as always. Anytime you'd want to infer something requires quite a heavy load, which is the main reason I haven't pushed forward.
> 
> @vedantroy keep us updated if you test this out!

yes, this is exactly right

within google, they all use this t5x library

**Comment by vedantroy** [comment]
> The encoder is frozen, not trained at all, and the text embeddings can even be pre-computed. So the size of text encoder should have negligible impact in the training cost.

That's something I've been thinking about. I've been looking into pre-computing the embeddings, and while it's possible it takes up a ton of space. So using something like GPT-RETRO might be a good alternative.

**Comment by vedantroy** [comment]
I have some general questions that I'd be curious to discuss with people here. Namely, 

- How much does shrinking the output space affect training efficiency? I've been trying to train a model that only generates cartoons, however, I'm not actually sure if shrinking the image space actually helps.
- Can we modify the space of captions and still get good results? It's hard to generate captions for images, but..., it's easy to (for example) generate tags for images. Is feeding a string of concatenated tags to T5 going to still be effective? How do we determine how "well" a caption is being processed by T5?
- Probably the biggest one: **How much does data quality matter**. Looking at datasets like LAION-5B, it doesn't seem like the (image, caption) quality is not very high, but it seems like this might just not matter??

**Comment by jorgemcgomes** [comment]
> * Can we modify the space of captions and still get good results? It's hard to generate captions for images, but..., it's easy to (for example) generate tags for images. Is feeding a string of concatenated tags to T5 going to still be effective? How do we determine how "well" a caption is being processed by T5?

A list of labels as input should still work. I've trained image generator models with lists of keywords with success, using T5 as encoder. And concatenated lists of labels is actually a part of the training data in Parti: 
"[...] JFT-4B dataset [44], which has images with text annotation labels. For textual descriptions of JFT, we randomly switch between the original labels as text (concatenated if an image has multiple labels) or machine-generated captions from a SimVLM model [...]" https://arxiv.org/pdf/2206.10789.pdf

What you need to consider is that, if the training data is **only** lists of labels, than there might be little benefit in increasing the encoder size, since it's mostly going to be a bag-of-words model, there's little for the transformer to transform given that the context of each word should be mostly irrelevant.

**Comment by vedantroy** [comment]
> > * Can we modify the space of captions and still get good results? It's hard to generate captions for images, but..., it's easy to (for example) generate tags for images. Is feeding a string of concatenated tags to T5 going to still be effective? How do we determine how "well" a caption is being processed by T5?
> 
> A list of labels as input should still work. I've trained T5 models with lists of keywords as inputs, and works fine. And concatenated lists of labels is actually a part of the training data in Parti: "[...] _JFT-4B dataset [44], which has images with text annotation labels. For textual descriptions of JFT, we randomly switch between the original labels as text (concatenated if an image has multiple labels) or machine-generated captions from a SimVLM model_ [...]" https://arxiv.org/pdf/2206.10789.pdf
> 
> What you need to consider is that, if the training data is **only** lists of labels, then there might be little benefit in increasing the encoder size, since it's mostly going to be a bag-of-words model, there's little for the transformer to transform given that the context of each word should be mostly irrelevant.

Wow, it is very interesting that PARTI was trained on labels / machine generated captions. This seems to indicate that human labelling is completely unnecessary to get data.

It's interesting that you can train only on tags though. Because the Imagen paper indicates that scaling language model size really helps, and it seems like using tags negates the benefits of language models.
